universum/configuration_support.py

```diff
@@ -1,18 +1,210 @@
+        self.if_env_set: str = if_env_set
+        self.children: Optional['Variations'] = None
+        self.extras: Dict[str, str] = extras if extras else {}
+        for key, value in kwargs.items():
+            self.extras[key] = value
+
+    def __repr__(self) -> str:
+        """
+        This function simulates dict-like legacy representation for output
+        :return: dict-like string
+
+        >>> cfg = ProjectConfiguration(name='foo', command=['bar'], my_var='baz')
+        >>> repr(cfg)
+        "{'name': 'foo', 'command': 'bar', 'my_var': 'baz'}"
+        """
+        res = {k: v for k, v in self.__dict__.items() if v and k != 'extras'}
+        res.update(self.extras)
+        if len(self.command) == 1:  # command should be printed as one string, instead of list
+            res['command'] = self.command[0]
+        return str(res)
+
+    def __eq__(self, other: Any) -> bool:
+        """
+        This functions simulates dict-like legacy comparison
+
+        :param other: dict to compare values, or object to check exact equality
+        :return: 'True' if 'other' matches
+
+        >>> cfg1 = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg2 = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg3 = ProjectConfiguration(name='foo', my_var='bar', critical=True)
+        >>> cfg1 == cfg1
+        True
+        >>> cfg1 == cfg2
+        True
+        >>> cfg1 == cfg3
+        False
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar'}
+        True
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar', 'critical': False}
+        True
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar', 'test': None}
+        True
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar', 'test': ' '}
+        False
+        """
+        if isinstance(other, ProjectConfiguration):
+            return self == other.__dict__
+        if isinstance(other, dict):
+            for key, val in other.items():
+                if val and self[key] != val:
+                    return False
+            return True
+        return super().__eq__(other)
+
+    def __getitem__(self, item: str) -> Optional[str]:
+        """
+        This functions simulates dict-like legacy access
+
+        :param item: client-defined item
+        :return: client-defined value
+
+        >>> cfg = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg['name']
+        'foo'
+        >>> cfg['my_var']
+        'bar'
+        >>> cfg['test']
+        """
+        return self.extras.get(item, self.__dict__.get(item, None))
+
+    def __add__(self, other: 'ProjectConfiguration') -> 'ProjectConfiguration':
+        """
+        This functions defines operator ``+`` for :class:`.ProjectConfiguration` class objects by
+        concatenating strings and contents of dictionaries
+
+        :param other: `ProjectConfiguration` object
+        :return: new `ProjectConfiguration` object, including all attributes from both `self` and `other` objects
```

Comment:
Please mention that "critical" attribute is skipped.

Reply:
should it really be skipped btw?

Reply:
The answer is in universum core logic.
The behavior of critical is different depending on where it is set to True.
If it is set in parent, but not in children, all children will be executed, but if any of them fails, the entire execution will stop after that. If critical is set on parent and children, the children will no be executed as soon as one of them fails.

Because of that, in contrast to all other attributes the propagation from parent to child does not happen.